actors/miner/tests/miner_actor_test_wpost.rs

```diff
@@ -90,7 +92,390 @@ fn basic_post_and_dispute() {
 }
 
 #[test]
-fn invalid_submissions() {}
+fn invalid_submissions() {
+    let period_offset = ChainEpoch::from(100);
+    let precommit_epoch = ChainEpoch::from(1);
+
+    let mut h = ActorHarness::new(period_offset);
+    let mut rt = h.new_runtime();
+    rt.epoch = precommit_epoch;
+    rt.balance.replace(TokenAmount::from(BIG_BALANCE));
+
+    h.construct_and_verify(&mut rt);
+
+    let sectors = h.commit_and_prove_sectors(&mut rt, 1, DEFAULT_SECTOR_EXPIRATION, vec![], true);
+    let sector = sectors[0].clone();
+    let pwr = miner::power_for_sector(h.sector_size, &sector);
+
+    // Skip to the due deadline.
+    let state = h.get_state(&rt);
+    let (dlidx, pidx) = state.find_sector(&rt.policy, &rt.store, sector.sector_number).unwrap();
+    let dlinfo = h.advance_to_deadline(&mut rt, dlidx);
+
+    // Invalid deadline.
+    {
+        let partition = miner::PoStPartition { index: pidx, skipped: make_empty_bitfield() };
+        let params = miner::SubmitWindowedPoStParams {
+            deadline: rt.policy.wpost_period_deadlines,
+            partitions: vec![partition],
+            proofs: make_post_proofs(h.window_post_proof_type),
+            chain_commit_epoch: dlinfo.challenge,
+            chain_commit_rand: Randomness(Vec::from(b"chaincommitment".clone())),
+        };
+        let result = h.submit_window_post_raw(
+            &mut rt,
+            &dlinfo,
+            vec![sector.clone()],
+            params,
+            PoStConfig::empty(),
+        );
+        expect_abort_contains_message(ExitCode::ErrIllegalArgument, "invalid deadline", result);
+        rt.reset();
+    }
+
+    // No partitions.
+    // This is a weird message because we don't check this precondition explicitly.
+    {
+        let params = miner::SubmitWindowedPoStParams {
+            deadline: dlinfo.index,
+            partitions: Vec::new(),
+            proofs: make_post_proofs(h.window_post_proof_type),
+            chain_commit_epoch: dlinfo.challenge,
+            chain_commit_rand: Randomness(Vec::from(b"chaincommitment".clone())),
+        };
+        let result = h.submit_window_post_raw(
+            &mut rt,
+            &dlinfo,
+            vec![sector.clone()],
+            params,
+            PoStConfig::empty(),
+        );
+        expect_abort_contains_message(
+            ExitCode::ErrIllegalArgument,
+            "expected proof to be smaller",
+            result,
+        );
+        rt.reset();
+    }
+
+    // Too many partitions.
+    {
+        let too_many = 11;
+        let mut partitions = Vec::with_capacity(too_many);
+        for i in 0..too_many {
+            let partition =
+                miner::PoStPartition { index: pidx + i as u64, skipped: make_empty_bitfield() };
+            partitions.push(partition);
+        }
```

Comment:
```suggestion
        let partitions: Vec<_> = (0u64..too_many).map(|i| miner::PoStPartition {
            index: pidx + i,
            skipped: make_empty_bitfield(),
        }).collect();
```

Reply:
that really isn't much more readable (and it is not more compact either).

Reply:
I find it easier to read and reason about. The reason is that `for` loops are often used in cases when you cannot use map, e.g. when having side-effects or you stop early. For me, the `map()` clearly indicates that you transform the whole range into something else.

Reply:
well fine, you have declared war on for+push, like you frown on golang.

I will grudgingly accept your suggestion, but note that I personally like the go approach to programming better; the less magic, the less cognitive overhead.